### PR TITLE
Driver: Refactor Hmd and Controller

### PR DESCRIPTION
**Pull request by Vixea**

### CODE REVIEW

File: alvr/server/cpp/alvr_server/alvr_server.cpp [hunk 1]
```diff
@@ -64,28 +64,48 @@ class DriverProvider : public vr::IServerTrackedDeviceProvider {
   public:
     std::shared_ptr<AdaptiveBitrate> adaptive_bitrate;
 
-    std::shared_ptr<OvrHmd> hmd;
-    std::shared_ptr<OvrController> left_controller, right_controller;
-    // std::vector<OvrViveTrackerProxy> generic_trackers;
+    std::shared_ptr<Hmd> hmd;
+    std::shared_ptr<Controller> controller;
+    std::shared_ptr<Controller> left_controller, right_controller;
```
Comment (deiteris) [inline alvr/server/cpp/alvr_server/alvr_server.cpp:69]: All these pointers aren't not shared. Just use `unique_ptr`.

File: alvr/server/cpp/alvr_server/alvr_server.cpp [hunk 2]
```diff
@@ -64,28 +64,48 @@ class DriverProvider : public vr::IServerTrackedDeviceProvider {
...
+        this->right_controller = std::make_shared<Controller>(RIGHT_HAND_ID);
 
         this->tracked_devices.insert({HEAD_ID, (TrackedDevice *)&*this->hmd});
-        if (this->left_controller && this->right_controller) {
-            this->tracked_devices.insert({LEFT_HAND_ID, (TrackedDevice *)&*this->left_controller});
-            this->tracked_devices.insert(
-                {RIGHT_HAND_ID, (TrackedDevice *)&*this->right_controller});
-        }
+            if (this->left_controller && this->right_controller) {
+                this->tracked_devices.insert({LEFT_HAND_ID, (TrackedDevice *)&*this->left_controller});
```
Comment (deiteris) [inline alvr/server/cpp/alvr_server/alvr_server.cpp:87]:
```suggestion
                this->tracked_devices.insert({LEFT_HAND_ID, (TrackedDevice *)this->left_controller.get()});
```

File: alvr/server/cpp/alvr_server/alvr_server.cpp [hunk 3]
```diff
@@ -64,28 +64,48 @@ class DriverProvider : public vr::IServerTrackedDeviceProvider {
...
+                this->tracked_devices.insert(
+                    {RIGHT_HAND_ID, (TrackedDevice *)&*this->right_controller});
```
Comment (deiteris) [inline alvr/server/cpp/alvr_server/alvr_server.cpp:89]:
```suggestion
                    {RIGHT_HAND_ID, (TrackedDevice *)this->right_controller.get()});
```

File: alvr/server/cpp/alvr_server/alvr_server.cpp [hunk 4]
```diff
@@ -64,28 +64,48 @@ class DriverProvider : public vr::IServerTrackedDeviceProvider {
   public:
     std::shared_ptr<AdaptiveBitrate> adaptive_bitrate;
 
-    std::shared_ptr<OvrHmd> hmd;
-    std::shared_ptr<OvrController> left_controller, right_controller;
-    // std::vector<OvrViveTrackerProxy> generic_trackers;
+    std::shared_ptr<Hmd> hmd;
+    std::shared_ptr<Controller> controller;
```
Comment (deiteris) [inline alvr/server/cpp/alvr_server/alvr_server.cpp:68]: Seems to be unused?

File: alvr/server/cpp/alvr_server/alvr_server.cpp [hunk 5]
```diff
@@ -64,28 +64,48 @@ class DriverProvider : public vr::IServerTrackedDeviceProvider {
...
+                    left_controller->GetSerialNumber().c_str(),
+                    vr::TrackedDeviceClass_Controller, //left_controller->getControllerDeviceClass(), can't solve linking error so leaving it like this for now
+                    left_controller.get());
+                if (!ret) {
+                    Warn("Failed to register left controller");
+                }
+
+                ret = vr::VRServerDriverHost()->TrackedDeviceAdded(
+                    right_controller->GetSerialNumber().c_str(),
+                    vr::TrackedDeviceClass_Controller, //right_controller->getControllerDeviceClass(), can't solve linking error so leaving it like this for now
```
Comment (deiteris) [inline alvr/server/cpp/alvr_server/alvr_server.cpp:103]: The issue with linker is that this is not member of the class. You added a method prototype `Controller.h`, tried to reference, but there's no implementation in cpp.

**Review by zarik5 — CHANGES_REQUESTED** [review]
Looks good, but please format the files you modified (i just run alt+shift+F on vscode).

**Review by zarik5 — APPROVED** [review]

**Pull request merged**